cpp/src_prims/linalg/lstsq.cuh

```diff
@@ -21,74 +21,316 @@
+                            // it's destroyed upon exit of the routine.
+                const int n_rows,
+                const int n_cols,
+                const math_t* b,
+                math_t* w,
+                cudaStream_t stream)
 {
+  const int minmn              = min(n_rows, n_cols);
   cusolverDnHandle_t cusolverH = handle.get_cusolver_dn_handle();
-  cublasHandle_t cublasH       = handle.get_cublas_handle();
-
-  ASSERT(n_rows > 1, "lstsq: number of rows cannot be less than two");
+  int cusolverWorkSetSize      = 0;
+  CUSOLVER_CHECK(raft::linalg::cusolverDngesvd_bufferSize<math_t>(
+    cusolverH, n_rows, n_cols, &cusolverWorkSetSize));
 
-  size_t U_len = n_rows * n_cols;
-  size_t V_len = n_cols * n_cols;
+  rmm::device_uvector<math_t> workset(cusolverWorkSetSize  // cuSolver
+                                        + n_rows * minmn   // U
+                                        + n_cols * n_cols  // V
+                                        + minmn            // S
+                                        + minmn            // U^T * b
+                                        + 1                // devInfo
+                                      ,
+                                      stream);
+  math_t* cusolverWorkSet = workset.data();
+  math_t* U               = cusolverWorkSet + cusolverWorkSetSize;
+  math_t* Vt              = U + n_rows * minmn;
+  math_t* S               = Vt + n_cols * n_cols;
+  math_t* Ub              = S + minmn;
+  int* devInfo            = reinterpret_cast<int*>(Ub + minmn);
 
-  rmm::device_uvector<math_t> S(n_cols, stream);
-  rmm::device_uvector<math_t> V(V_len, stream);
-  rmm::device_uvector<math_t> U(U_len, stream);
+  CUSOLVER_CHECK(raft::linalg::cusolverDngesvd<math_t>(cusolverH,
+                                                       'S',
+                                                       'S',
+                                                       n_rows,
+                                                       n_cols,
+                                                       A,
+                                                       n_rows,
+                                                       S,
+                                                       U,
+                                                       n_rows,
+                                                       Vt,
+                                                       n_cols,
+                                                       cusolverWorkSet,
+                                                       cusolverWorkSetSize,
+                                                       nullptr,
+                                                       devInfo,
+                                                       stream));
+  raft::linalg::gemv(handle, U, n_rows, minmn, b, Ub, true, stream);
+  auto f = [] __device__(math_t a, math_t b) { return raft::myAbs(b) > math_t(1e-10) ? a / b : a; };
```

Comment:
I feel silly to admit it, but I tried to define it as a plain `__device__` function within the file and constantly `cudaErrorIllegalAddress` :) Do you have an idea why this could happen? (I tried both with/without inline or forceinline modifiers)

Reply:
Can we instead define a struct functor and make the operator function a device function?

Reply:
That works, thanks!